packages/rum-core/test/common/page-visibility.spec.js

```diff
@@ -0,0 +1,299 @@
+
+          unobservePageVisibility = observePageVisibility(
+            configService,
+            transactionService
+          )
+          hidePageSynthetically(eventName)
+
+          expect(state.lastHiddenStart).toBe(anyTime)
+        })
+      })
+
+      describe(`with a transaction that has not ended yet`, () => {
+        // Starts performanceMonitoring to observe how transaction ends
+        function startsPerformanceMonitoring() {
+          const performanceMonitoring = serviceFactory.getService(
+            'PerformanceMonitoring'
+          )
+          performanceMonitoring.init()
+        }
+
+        function createTransaction() {
+          const transaction = transactionService.startTransaction(
+            'test-tr',
+            'test-type'
+          )
+
+          let span = transaction.startSpan('test span', 'test span type', {
+            startTime: 0
+          })
+          span.end(100)
+
+          return transaction
+        }
+
+        it('should end the transaction', () => {
+          const transaction = createTransaction()
+          const endTransactionSpy = spyOn(transaction, 'end')
+          spyOn(transactionService, 'getCurrentTransaction').and.returnValue(
+            transaction
+          )
+
+          unobservePageVisibility = observePageVisibility(
+            configService,
+            transactionService
+          )
+          hidePageSynthetically(eventName)
+
+          expect(endTransactionSpy).toHaveBeenCalledTimes(1)
+        })
+
+        it('should dispatch the QUEUE_FLUSH event once the transaction ends and added to the queue', async () => {
+          // Arrange
+          const transaction = createTransaction()
+          const dispatchEventSpy = spyOn(
+            configService,
+            'dispatchEvent'
+          ).and.callThrough()
+          spyOn(transaction, 'end').and.callThrough()
+          spyOn(transactionService, 'getCurrentTransaction').and.returnValue(
+            transaction
+          )
+          startsPerformanceMonitoring()
+
+          // Act
+          unobservePageVisibility = observePageVisibility(
+            configService,
+            transactionService
+          )
+          hidePageSynthetically(eventName)
+          await waitFor(() => dispatchEventSpy.calls.any())
+
+          // Assert
+          expect(dispatchEventSpy).toHaveBeenCalledWith(QUEUE_FLUSH)
+        })
+
+        it('should remove the QUEUE_ADD_TRANSACTION event listener once the transaction ends and added to the queue', async () => {
+          // Arrange
+          const unobserveSpy = jasmine.createSpy()
+          const observeEvent = configService.observeEvent.bind(configService)
+          spyOn(configService, 'observeEvent').and.callFake((name, fn) => {
+            observeEvent(name, fn)
+            return unobserveSpy
+          })
+          const transaction = createTransaction()
+          const dispatchEventSpy = spyOn(
```

Comment:
It is true that is a bit cumbersome. It is not enough because these tests (more "integration" than unit ones) we are reproducing the following scenario:

1. observe page visibility
2. hide page to execute the logic that ends an ongoing transaction
3. The transaction end method will dispatch the `QUEUE_ADD_TRANSACTION` event behind the curtains
4. All what we want to assert takes places within the `QUEUE_ADD_TRANSACTION` event listener. In this particular test means to make sure that the `unobserveSpy` is being called too.

-- 
It is true too that another possible way is to stop relying on the integration of the different components and to call to dispatchEvent(QUEUE_ADD_TRANSACTION), but for this case I prefer the integration way as well.

P.S. I removed `spyOn(transaction, 'end').and.callThrough()` since was not required at all

Reply:
Gotcha, I ll leave it up to you then.